physicalplan/aggregate/hashaggregate.go

```diff
@@ -82,95 +83,76 @@ func (a *aggregate) Next(ctx context.Context) ([]model.StepVector, error) {
 	}
 	defer a.next.GetPool().PutVectors(in)
 
-	a.once.Do(func() { err = a.initOutputBuffers(ctx) })
+	a.once.Do(func() { err = a.initializeTables(ctx) })
 	if err != nil {
+		a.workers.Shutdown()
 		return nil, err
 	}
 
-	result := make([]model.StepVector, len(in))
+	result := a.vectorPool.GetVectorBatch()
 	for i, vector := range in {
 		a.workers[i].Send(vector)
 	}
 
 	for i, vector := range in {
 		a.workers[i].Done()
-		result[i] = a.workers[i].GetOutput()
-		a.next.GetPool().PutSamples(vector.Samples)
-
+		result = append(result, a.workers[i].GetOutput())
+		a.next.GetPool().PutStepVector(vector)
 	}
-	return result, nil
-}
 
-func (a *aggregate) shutdownWorkers() {
-	for i := 0; i < len(a.workers); i++ {
-		a.workers[i].Shutdown()
-	}
+	return result, nil
 }
 
-func (a *aggregate) initOutputBuffers(ctx context.Context) error {
+func (a *aggregate) initializeTables(ctx context.Context) error {
 	series, err := a.next.Series(ctx)
 	if err != nil {
 		return err
 	}
 
-	inputCache := make([]uint64, len(series))
-	outputMap := make(map[uint64]*aggregateResult)
-	outputCache := make([]*aggregateResult, 0)
+	if a.by && len(a.labels) == 0 {
+		tables, err := newVectorizedTables(a.stepsBatch, a.aggregation)
+		if err != nil {
+			return err
+		}
+		a.tables = tables
+		a.series = []labels.Labels{{}}
+		a.vectorPool.SetStepSize(1)
+		return nil
+	}
 
+	inputCache := make([]uint64, len(series))
+	outputMap := make(map[uint64]*model.Series)
+	outputCache := make([]*model.Series, 0)
 	for i := 0; i < len(series); i++ {
 		buf := make([]byte, 128)
```

Comment:
Can we move the buf out of the for loop? To reuse the buffer

Reply:
Nice catch, updated.